interfaces/builtin: create can-bus interface

jocave opened the pull request:
Create a snapd interface to allow communication on a CAN network on devices supporting [SocketCAN](https://www.kernel.org/doc/Documentation/networking/can.txt). 

The interface should allow bind of RAW sockets in family PF_CAN.

The interface also allows use of `ip` for configuration of the network interface.

I tested the interface using the vcan kernel module and a [CANable](http://canable.io/) USB CAN device. I checked I could perform some simple configuration actions e.g. for vcan:
```
ip link add vcan0 type vcan
ip link set vcan0 up
```
and on CANable:
```
ip link set can0 type can bitrate 1000000
ip link set can0 up
```
I then created a test snap of `can-utils` and checked that `candump` and `cansend` were able to exchange data when the `socketcan` plug was connected.

zyga commented on interfaces/builtin/socketcan.go:
```diff
@@ -0,0 +1,61 @@
+const socketCanConnectedPlugAppArmor = `
+# Description: Can access SocketCAN networks.
+network can,
+network netlink raw,
```
Can you please provide the rationale for using netlink sockets? A quick google on wikipedia seems to imply this is not required (see: https://en.wikipedia.org/wiki/SocketCAN)

zyga commented on interfaces/builtin/socketcan.go:
```diff
@@ -0,0 +1,61 @@
+network can,
+network netlink raw,
+
+capability net_admin,
```
Is this required to use CAN sockets or is this associated with netlink somehow?

Do you have a test program that we can use to experiment with this?

zyga commented on interfaces/builtin/socketcan.go:
```diff
@@ -0,0 +1,61 @@
+capability net_admin,
+
+# Allow configuration of the interface using ip
+/{,usr/}{,s}bin/ip ixr,
```
This seems to imply there would be a canbus-control and regular canbus interface (without the ability to configure can).

jdstrand replied:
I can answer @zyga's questions. The ip command will require net_admin for setting up the interface and also needs netlink raw for some operations. Allow of this is covered by network-control, but it seems useful to not lump SocketCAN into all other networking since it is not so much about network connectivity in the traditional sense as about using the socket() API in the kernel.

As such, please adjust your rules to be:
```
# Description: Can configure and use SocketCAN networks
network can,

# Allow configuration of the interface using the ip command for SocketCAN
/{,usr/}{,s}bin/ip ixr,

# required by ip to configure SocketCAN networks
capability net_admin,
network netlink raw,
```

jdstrand commented on interfaces/builtin/socketcan.go:
```diff
@@ -0,0 +1,61 @@
+
+socket AF_NETLINK - NETLINK_ROUTE
+socket AF_NETLINK - NETLINK_GENERIC
+`
```
All of these are in network-control.

jdstrand commented on interfaces/builtin/socketcan.go:
```diff
@@ -0,0 +1,61 @@
+bind
+
+socket AF_NETLINK - NETLINK_ROUTE
+socket AF_NETLINK - NETLINK_GENERIC
```
Please add a comment above these:
```
# required by ip to configure SocketCAN networks
```

zyga commented:
@jdstrand I'm +1 on making this socketcan-control and I'd suggest bundling socketcan into network/network-bind. I don't see a special need for socketcan if we are about to auto-connect it (which I think makes sense once socketcan-control is separate).

jocave commented:
@zyga @jdstrand making this socketcan-control sounds fine to me, I'll make those changes.

For me this point rings true:
>  it is not so much about network connectivity in the traditional sense as about using the socket() API in the kernel

These are two different network types by physically and in conceptual usage so for me bundling access to CAN in to the network interfaces doesn't make sense (allow access to one doesn't imply you want to allow access to the other). 

Examples of the types of device parameters "controlled" using `ip` are: interface being up/down, bitrate, MTU size i.e. FD mode. Note that CAN is a "broadcast only" network, so there is no identification of recipient and identification of sender is performed in the preamble of the packet (and is also an indication of priority). All nodes on the CAN network must use the same bitrate - this is important in this context because I think it means you might trust a snap to send packets, but not modify bitrate and potentially prevent another application on the device from communicating. Hence I think a non-control socketcan interface is a good idea.

jdstrand commented:
@zyga (cc @niemeyer) - I agree with @jocave that this is sufficiently different from network, network-control and friends. As I said in my review: "All of this is covered by network-control, but it seems useful to not lump SocketCAN into all other networking since it is not so much about network connectivity in the traditional sense as about using the socket() API in the kernel." Put another way, the fact that SocketCan is using the socket() API is a developer convenience, much like Unix Domain Sockets. Furthermore, I was trying to imagine consumers of this interface (eg, automotive, automation) and it seemed liked a nice property that these applications would *not* need all the additional access that network-control and network provide.

jocave commented:
The commits above do the following in order:

- make the changes to the policies requested by @jdstrand 
- rename the interface to `socketcan-control`
- in lieu of more opinions, created a non-control `socketcan` interface

I created a snap of `can-utils` to test the two interfaces and have pushed it [here](https://git.launchpad.net/~jocave/+git/can-utils/tree/?h=interface-test-snap)

You should be able to try the `vcan` apps on any ubuntu machine by first doing a `sudo modprobe vcan`. HTH

jdstrand commented:
The tests are failing because of:
Missing high-level test for interface 'socketcan'. Please add to:
* tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml
* tests/lib/snaps/test-snapd-policy-app-provider-core/meta/snap.yaml (if needed)
* tests/lib/snaps/test-snapd-policy-app-provider-classic/meta/snap.yaml (if needed)

jdstrand commented on interfaces/builtin/socketcan.go:
```diff
@@ -0,0 +1,53 @@
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
```
I feel like your instinct here to disallow `socketcan` is correct, but for completeness of the review, why should the use of CAN networking be denied by default? (Again, I think I agree with you when imagining snaps in automotive/etc)

jocave replied:
Given CAN is often used in industrial and automotive applications (e.g. passenger lift controls, driving assistance systems) and these could have a role in ensuring human safety, I feel it is better to be conservative and allow store owners or device administrators to grant access to the resource in this case.

jdstrand commented on interfaces/builtin/socketcan.go:
```diff
@@ -0,0 +1,53 @@
+		name:                  "socketcan",
+		summary:               socketCanSummary,
+		implicitOnCore:        true,
+		implicitOnClassic:     true,
```
We can't do it only on gadgets until we have dynamic slots (which are currently being worked on). Note that CONFIG_CAN is enabled in the Ubuntu 18.04 LTS kernel, so it is technically available on classic. I think it is fine to expose implicitly both places (particularly since it is manually connected).

jdstrand commented on interfaces/builtin/socketcan.go:
```diff
@@ -0,0 +1,53 @@
+`
+
+const socketCanConnectedPlugAppArmor = `
+# Description: Can access SocketCAN networks
```
For symmetry with the other interface, perhaps "Can use SocketCAN networking"?

jdstrand commented on interfaces/builtin/socketcan_control.go:
```diff
@@ -0,0 +1,64 @@
+`
+
+const socketCanControlConnectedPlugAppArmor = `
+# Description: Can configure and use SocketCAN networks
```
s/networks/networking/

jdstrand commented on interfaces/builtin/socketcan_control_test.go:
```diff
@@ -0,0 +1,111 @@
+
+func (s *SocketCanControlInterfaceSuite) TestAutoConnect(c *C) {
+	c.Assert(s.iface.AutoConnect(s.plugInfo, s.slotInfo), Equals, true)
+}
```
These days TestAutoConnect can be omitted (this is tested via interfaces/policy)

jdstrand commented on interfaces/builtin/socketcan_test.go:
```diff
@@ -0,0 +1,110 @@
+
+func (s *SocketCanInterfaceSuite) TestAutoConnect(c *C) {
+	c.Assert(s.iface.AutoConnect(s.plugInfo, s.slotInfo), Equals, true)
+}
```
ditto

jdstrand commented on tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml:
```diff
@@ -263,6 +263,9 @@ apps:
     plugs: [ daemon-notify ]
+  socketcan-control:
+    command: bin/run
+    plugs: [ socketcan-control ]
```
Need to add the one for 'socketcan' then travis should start passing.

mvo5 commented:
Looks fine, I think we just need to get an approval from @niemeyer on the name. A spread test would be nice but I understand this is tricky because the can socket is not available in every kernel.

niemeyer commented on interfaces/builtin/socketcan_control.go:
```diff
@@ -0,0 +1,64 @@
+
+# required by ip to configure SocketCAN networks
+capability net_admin,
+network netlink raw,
```
Isn't that opening up to administration of non-CAN interfaces too?

zyga replied:
Yes, apparently this is required by the can management tools.

niemeyer commented on interfaces/builtin/socketcan_control.go:
```diff
@@ -0,0 +1,64 @@
+
+# required by ip to configure SocketCAN networks
+socket AF_NETLINK - NETLINK_ROUTE
+socket AF_NETLINK - NETLINK_GENERIC
```
Similarly, these are not specific to CAN.

The split here seems a bit suspect. It seems to grant the same thing that the use of can, above, plus the configuration of the network in general.

Can we please have a call tomorrow? I may be completely off, but would like to at least understand this a bit better if you could spare a moment.

jdstrand replied:
The reasoning behind both of this is captured in these comments:
* https://github.com/snapcore/snapd/pull/5279#issuecomment-395735128
* https://github.com/snapcore/snapd/pull/5279#issuecomment-395738522

Basically, yes, there is overlap between can-bus/network and can-bus-control/network-control (net_admin is in a lot of ways a catch all, like sys_admin-- the Linux capabilities are unfortunately not finegrained). However, network-control is *big* and we felt that something that only wanted to administer can-bus didn't need all of network-control. Specifically, it doesn't need to change hostname, bridges, inet, inet6, firewalling, ieee80211, rfkill, resolvconf, tun/tap, network namespaces, sys_admin, etc. It only needs 'ip', net_admin, and AF_CAN. These NETLINK_ROUTE and NETLINK_GENERIC are similarly not CAN-specific, but required by the ip command. Put another way, we broke it out since CAN devices might play a role in human safety and limiting what both can-bus and can-bus-control interfaces could do was deemed a good thing.

niemeyer commented on interfaces/builtin/socketcan.go:
```diff
@@ -0,0 +1,53 @@
+
+package builtin
+
+const socketCanSummary = `allows use of SocketCAN network interfaces`
```
This would be equivalent to saying something like "allows use of the socket berkley interface" for the network interface, but I think we want to highlight what we're granting access to instead of the API being used for it. Perhaps something like:
```
`allows access to the CAN bus`
```

niemeyer commented on interfaces/builtin/socketcan_control.go:
```diff
@@ -0,0 +1,64 @@
+`
+
+const socketCanControlConnectedPlugSecComp = `
+# Description: Can use SocketCAN networking
```
use vs. control

niemeyer commented on interfaces/builtin/socketcan_control.go:
```diff
@@ -0,0 +1,64 @@
+
+package builtin
+
+const socketCanControlSummary = `allows configuration and use of SocketCAN network interfaces`
```
`allows configuration of CAN network interfaces`

jocave commented:
I feel, in this case, that what we are granting applications access to is specifically a hardware device and associated driver that supports SocketCAN operation. 

There are other methods of interaction with CAN Bus chipsets that this interface (as it stands) does not attempt to address e.g. can4linux, slcand etc. These can imply file type operations as opposed to socket operations and therefore quite different behaviour by the application.

Hence my preference would be to set expectations by keeping the name as _socketcan_ and keeping the access provided by the interface limited to this context.

jocave commented:
I'll rework based on the above and check with my test applications again.

jocave commented:
The above two commits are an attempt to implement the changes discussed.

I made these changes are then tested using the hardware I have available. I modified my test `can-utils` snap to use the `can-bus` and `network-control` interfaces and have pushed this [here](https://git.launchpad.net/~jocave/+git/can-utils/tree/snap/snapcraft.yaml?h=network-control-test-snap). Everything worked as expected.

I changed the summary text to match the suggestion from one of the earlier comments. Note that I did leave the description comment in the snippets as "Can use CAN networking" given that for now this is the supported method of interaction.

Please give me your feedback on this revised implementation. Thanks

jdstrand commented on interfaces/builtin/can_bus.go:
```diff
@@ -0,0 +1,53 @@
+
+const canBusConnectedPlugSecComp = `
+# Description: Can use CAN networking
+bind
```
Can you add this comment so it's clear to can-bus consumers:
```
# We allow AF_CAN in the default template since it is mediated via the AppArmor rule
#socket AF_CAN
```

jocave replied:
Done
